### ISSUE TITLE
Activation behavior of Components on referencing without injection.

### ISSUE THREAD

**Issue by juergen-albert** [osgi/osgi#420]
I have a bit of an odd Situation and I'm not sure if I have a faulty
expectation:

I have 2 Components:

```java
@Component(immediate = true, property = Condition.CONDITION_ID + "=" +
CONDITION_EQUINOX_CONFIG)
public class EquinoxConfigInitializerImpl implements Condition {
     @Activate
     public void activate(ComponentContext ctx) {
         // some code here
     }
}
```

I have a second component that does require this Condition to be there
and to be activated. It does not need to be injected, so I reference it
only in the annotation.

```java
@Component(immediate=true , reference = {
     @Reference(name = "equinoxConfig", service = Condition.class,
policy=ReferencePolicy.STATIC,
cardinality=ReferenceCardinality.MANDATORY, target = "(" +
Condition.CONDITION_ID + "=" + CONDITION_EQUINOX_CONFIG + ")")
} )

public class Component2 { ... }
```

My expectation thus has been, that the EquinoxConfigInitializerImpl
would have been activated before Component2. For some reason this is not
the case and it is activated after Component2. I couldn't really find
something concrete in the Spec that lead me to a definitive answer.

Am I wrong, is this undefined or did I find a Bug in SCR (2.1.16)?

Answer @bjhargrave :

 It is possible that the ServiceRegistration for the immediate component happens before the activation of the immediate component. The DS spec does not require that an immediate component be activated before its service is registered. It only requires the immediate component to be activated once all dependencies are satisfied.

I do not see an error in the SCR implementation nor a problem in the DS spec. Without injection, you don't cause a happens-before relationship between the activation of the referenced (immediate) component and the activation of the referencing component. Since the referencing component does not actually cause the immediate component to be injected, the referencing component activation can proceed before the immediate component is activated. The activation of the 2 components is unordered.

I think the simple solution is to just use field or constructor injection in the referencing component to the referenced component. This will establish the desired ordering.

**Comment by juergen-albert** [comment]
I believe this warrants some further discussion. We should either describe this behavior in the Spec or think about it. The difference in activation time might not be what the user expects, especially when a Service is Immediate.

**Comment by tjwatson** [comment]
See https://issues.apache.org/jira/browse/FELIX-5570 for some details.  David Jencks had indicated that this behavior was mandated by the specification

```
Felix DS used to create the component instance before registering the service, but the spec explicitly says to register the service first. IIRC there were a bunch of deadlocks that occurred with either order that had to be solved, but I don't recall the details.
```

So 1) we need to confirm what David says about the spec requirement, 2) I don't look forward to solving a whole new set of deadlock issues this may expose if we are mandated to create immediate components before registration.

**Comment by rotty3000** [comment]
@juergen-albert this is the same behaviour you will see if your `@Reference` field | constructor/method parameter is either `ServiceReference<S>` or `Map<String, Object>` so I believe there is much precedent within the spec itself for how things work today.

**Comment by bjhargrave** [comment]
Service registration happens-before activation. This enables delayed components. An immediate component is activate once it is satisfied. Satisfaction is not coupled to service registration. Immediate components do not need to be services. 

The spec must not require immediate component activation to happen-before service registration.

In your example, I think a case could be made that the referenced `Condition` service is a bound service of the second component and that SCR must ensure that all bound services are activated before the referencing component is activated. SCR would need to ensure that any bound service that was injected via field, method, or constructor injection was activated. So why not a bound service that is not injected?

As a counter to that position, the `reference` element of the `Component` annotation specifically states it is for "The lookup strategy references of this Component." Such references are obtained from a `locateService` method on `ComponentContext` and this would be the point at which SCR must activate the component instance.

So I do think a case could be made in either direction of the above but I do not agree that the spec must require immediate components to be activated before their service registration.

**Comment by juergen-albert** [comment]
> In your example, I think a case could be made that the referenced `Condition` service is a bound service of the second component and that SCR must ensure that all bound services are activated before the referencing component is activated. SCR would need to ensure that any bound service that was injected via field, method, or constructor injection was activated. So why not a bound service that is not injected?

This sums up my intent nicely. If push comes to shove, I can create an injection and handle it like this. In the end, this would be useless code and may make sonarcube or other code quality tools complain and for good reason. Thinking bit more in the direction of the activation Condition for the DS, I ask myself if we would have the same behavior. If so, I would always need to register a Condition manually at the end of my component activation, in order to make sure that it states what it is for.

**Comment by bjhargrave** [comment]
> In the end, this would be useless code and may make sonarcube or other code quality tools complain and for good reason.

You can inject in a field or constructor parameter and tell sonarcube to shut up about it.



> Thinking bit more in the direction of the activation Condition for the DS, I ask myself if we would have the same behavior. If so, I would always need to register a Condition manually at the end of my component activation, in order to make sure that it states what it is for.

It is not clear from this but I assume you are talking about the satisfying condition.

Conditions are meant to be marker services. The act of service registration implies the presence of the condition. You seem to be wanting to combine the activation of the service object behind the Condition service with the presence of the condition when it is the other way around. First the state is set to meet the condition, then the Condition service is registered to indicate the condition is met. But DS does the opposite. DS first registers the service and then activates the object. So other non-DS users of the Condition service will, properly, assume the service registration is sufficient to know the condition is present and not also then get the service object.

So I think it is a bad practice to require those who depend upon a condition to depend upon more than the Condition service registration.

In your example, you really need to do the work in the `EquinoxConfigInitializerImpl.activate` _before_ the Condition service is registered. So even with a change to the DS spec as above, you are still out-of-order. For a condition, the condition state must be present _before_ you register the Condition service. It is the service registration that signals the condition not the obtaining of the service object behind the Condition service.

**Comment by bjhargrave** [comment]
Closing as there is a design mistake in the example code. No DS spec problems identified.

**Comment by juergen-albert** [comment]
As I had to leave the last call before this issue came up, I'd like to revisit this topic on our next call. 

I do understand, that this is a feature, not a bug. I was able to explain the behavior to myself as I realized what happens in real life. This is more an issue of expectations. I would see myself as a seasoned OSGi user and the reaction of myself and everybody I discussed the issue with, before I wrote to the mailing list, was surprise. Thus I would postulate that this might be a common misconception. 

I'd like to step back a bit from the fact that my example uses a `Condition` as trigger and dive in the history on how I arrived at this point. The Idea behind the [Condition Service Specification](http://docs.osgi.org/specification/osgi.core/8.0.0/service.condition.html) came from another Project. There we have a couple of indirect service requirements, which lead to exceptions on startup. DS retried activation again and it sorted itself out in the end. As such Exceptions are annoying and looked bad, I started adding this indirect requirements as Field Injection points which solved this Problem. 
This however needed to be done in a lot of services and I always had to suppress the unused warning. At a certain point I realized that I can add references to `Component` Annotation and I got rid of the unnecessary injections. As we still had quite a large list of `@Reference` on each service, the idea of a Condition marker Service came to me and I ended up with one condition that collects everything and I only have one `@Reference` on each service. It lead to the Condition Specification  and my remaining Reference can be replaced with the activation condition when I update SCR the next time.

So much for the History. With what I know now, I realize that the case I described above only works out of dumb luck. A few of the services I base my condition on, need to be activated before it can be seen as satisfied. So the whole thing will work until it doesn't. Long story short, the difference between an injection and a plain reference without injection is not really apparent. We use the same Annotation, which this indicates similar behavior. I recognize that this is not an everyday case, but in the spirit of user friendliness I believe we should give this some more thought. My argument that is that even I, relatively well versed in the DS Specification ran into this situation.

Thus I see the following options:

* Do nothing, as it can be solved by addition an injection point for References that need to have and activated Instance.
* Change the DS Spec to always track a Service Instance and not only a `ServiceReference` when `@Reference` is used in a `@Component` (Which would be a change of the default behavior and thus will need much more discussion and consideration).
* Add an attribute to the `@Reference` and its XML counterpart to tell it to track an instance. At runtime SCR already determines this information by looking the XML and at the injection point (is it a `ServiceReference` or a service instance or is it no injection at all). Thus it would only affect the Situation for a `@Reference` without an injection.
